nautobot_circuit_maintenance/handle_notifications/sources.py

```diff
@@ -88,25 +108,124 @@ def init(cls: Type[T], name: str) -> Type[T]:
+    """Abstract class that shares some methods and attributes accross email based sources."""
+
+    account: str
+    emails_to_fetch = []
+
+    def validate_providers(self, logger: Job, notification_source: NotificationSource, since_txt: str) -> bool:
+        """Method to validate that the NotificationSource has attached Providers.
+
+        Args:
+            logger (Job): Job to use its logger
+            notification_source (NotificationSource): Notification Source to validate providers
+            since_txt (str): Date string to be used to log
+
+        Returns:
+            bool: True if there are relevant providers attached or False otherwise
+        """
+        providers_with_email = []
+        providers_without_email = []
+        if not notification_source.providers.all():
+            logger.log_warning(
+                message=f"Skipping source '{notification_source.name}' because no providers were defined.",
+            )
+            return False
+
+        for provider in notification_source.providers.all():
+            cm_cf = CustomField.objects.get(name="emails_circuit_maintenances")
+            provider_emails = provider.get_custom_fields().get(cm_cf)
+            if provider_emails:
+                self.emails_to_fetch.extend(provider_emails.split(","))
+                providers_with_email.append(provider.name)
+            else:
+                providers_without_email.append(provider.name)
+
+        if providers_without_email:
+            logger.log_warning(
+                message=(
+                    f"Skipping {', '.join(providers_without_email)} because these providers have no email configured."
+                ),
+            )
+
+        if not providers_with_email:
+            logger.log_warning(
+                message=(
+                    f"Skipping Notification Source {notification_source.name} because none of the related providers "
+                    "have emails defined."
+                ),
+            )
+            return False
+
+        logger.log_info(
+            message=(
+                f"Retrieving notifications from {notification_source.name} for "
+                f"{', '.join(providers_with_email)} since {since_txt}"
+            ),
+        )
+
+        return True
+
+    @staticmethod
+    def extract_email_source(email_source: str) -> str:
+        """Method to get the sender email address."""
+        try:
+            email_source = re.search(r"\<([A-Za-z0-9_@.]+)\>", email_source).group(1)
+        except AttributeError:
+            try:
+                email_source = re.search(r"([A-Za-z0-9_@.]+)", email_source).group(1)
+            except AttributeError:
+                return ""
+        return email_source
+
+    @staticmethod
+    def extract_provider_data_type(email_source: str) -> Tuple[str, str, str]:
+        """Method to extract the provider data type based on the referenced email for the provider."""
```

Comment:
Perhaps add documentation of the returned tuple and what each element represents?